commit 4a77b7bcc74f8bb731aac85d37e542f19b7cbb6d
Author: Susan Franklin <sfranklin>
Date:   Wed Mar 18 19:43:19 2009 +0000

    Bug 262611 - [ui] Software without defined names do not match filters

1	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/AvailableIUPatternFilter.java
3	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionWizardPage.java
3	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/SelectableIUsPage.java
3	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/AvailableIUGroup.java
3	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/InstalledIUGroup.java
3	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/RevertProfilePage.java
29	3	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/viewers/IUComparator.java
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/AvailableIUPatternFilter.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/AvailableIUPatternFilter.java
index b386ebc..27fe4ae 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/AvailableIUPatternFilter.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/AvailableIUPatternFilter.java
@@ -95,7 +95,7 @@ public class AvailableIUPatternFilter extends PatternFilter {
 				if (text != null && wordMatches(text))
 					return true;
 			}
-			if (checkId) {
+			if (checkId || (checkName && text == null)) {
 				text = iu.getId();
 				if (wordMatches(text)) {
 					return true;
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionWizardPage.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionWizardPage.java
index b571b0c..96afcd5 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionWizardPage.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionWizardPage.java
@@ -105,7 +105,9 @@ public abstract class ResolutionWizardPage extends ProvisioningWizardPage {
 		});
 
 		// Filters and sorters before establishing content, so we don't refresh unnecessarily.
-		treeViewer.setComparator(new IUComparator(IUComparator.IU_NAME));
+		IUComparator comparator = new IUComparator(IUComparator.IU_NAME);
+		comparator.useColumnConfig(getColumnConfig());
+		treeViewer.setComparator(comparator);
 		treeViewer.setComparer(new ProvElementComparer());
 
 		contentProvider = new ProvElementContentProvider();
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/SelectableIUsPage.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/SelectableIUsPage.java
index c7d5d1f..e25d651 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/SelectableIUsPage.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/SelectableIUsPage.java
@@ -109,7 +109,9 @@ public class SelectableIUsPage extends ProvisioningWizardPage implements ISelect
 		});
 
 		// Filters and sorters before establishing content, so we don't refresh unnecessarily.
-		tableViewer.setComparator(new IUComparator(IUComparator.IU_NAME));
+		IUComparator comparator = new IUComparator(IUComparator.IU_NAME);
+		comparator.useColumnConfig(ProvUI.getIUColumnConfig());
+		tableViewer.setComparator(comparator);
 		tableViewer.setComparer(new ProvElementComparer());
 
 		contentProvider = new ProvElementContentProvider();
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/AvailableIUGroup.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/AvailableIUGroup.java
index 36a8b36..3e208e8 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/AvailableIUGroup.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/AvailableIUGroup.java
@@ -140,7 +140,9 @@ public class AvailableIUGroup extends StructuredIUGroup {
 		labelProvider.setToolTipProperty(IInstallableUnit.PROP_DESCRIPTION);
 
 		// Filters and sorters before establishing content, so we don't refresh unnecessarily.
-		availableIUViewer.setComparator(new IUComparator(IUComparator.IU_NAME));
+		IUComparator comparator = new IUComparator(IUComparator.IU_NAME);
+		comparator.useColumnConfig(getColumnConfig());
+		availableIUViewer.setComparator(comparator);
 		availableIUViewer.setComparer(new ProvElementComparer());
 
 		// Now the content provider.
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/InstalledIUGroup.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/InstalledIUGroup.java
index 15aeaf1..c8f3f75 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/InstalledIUGroup.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/InstalledIUGroup.java
@@ -59,7 +59,9 @@ public class InstalledIUGroup extends StructuredIUGroup {
 		TreeViewer installedIUViewer = new TreeViewer(parent, SWT.MULTI | SWT.FULL_SELECTION | SWT.H_SCROLL | SWT.V_SCROLL | SWT.BORDER);
 
 		// Filters and sorters before establishing content, so we don't refresh unnecessarily.
-		installedIUViewer.setComparator(new IUComparator(IUComparator.IU_NAME));
+		IUComparator comparator = new IUComparator(IUComparator.IU_NAME);
+		comparator.useColumnConfig(getColumnConfig());
+		installedIUViewer.setComparator(comparator);
 		installedIUViewer.setComparer(new ProvElementComparer());
 
 		// Now the content.
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/RevertProfilePage.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/RevertProfilePage.java
index 993e885..8208600 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/RevertProfilePage.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/dialogs/RevertProfilePage.java
@@ -140,7 +140,9 @@ public class RevertProfilePage extends InstallationPage implements ICopyable {
 		Label label = new Label(composite, SWT.NONE);
 		label.setText(ProvUIMessages.RevertDialog_ConfigContentsLabel);
 		configContentsViewer = new TreeViewer(composite, SWT.MULTI | SWT.FULL_SELECTION | SWT.H_SCROLL | SWT.V_SCROLL | SWT.BORDER);
-		configContentsViewer.setComparator(new IUComparator(IUComparator.IU_NAME));
+		IUComparator comparator = new IUComparator(IUComparator.IU_NAME);
+		comparator.useColumnConfig(ProvUI.getIUColumnConfig());
+		configContentsViewer.setComparator(comparator);
 		configContentsViewer.setComparer(new ProvElementComparer());
 		configContentsViewer.setContentProvider(new DeferredQueryContentProvider());
 
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/viewers/IUComparator.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/viewers/IUComparator.java
index 05b317a..98505209 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/viewers/IUComparator.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/viewers/IUComparator.java
@@ -20,9 +20,26 @@ public class IUComparator extends ViewerComparator {
 	public static final int IU_NAME = 0;
 	public static final int IU_ID = 1;
 	private int key;
+	private boolean showingId = false;
 
 	public IUComparator(int sortKey) {
 		this.key = sortKey;
+		showingId = sortKey == IU_ID;
+	}
+
+	/**
+	 * Use the specified column config to determine
+	 * whether the id should be used in lieu of an empty name
+	 * when sorting.
+	 * 
+	 * @param columnConfig
+	 */
+	public void useColumnConfig(IUColumnConfig[] columnConfig) {
+		for (int i = 0; i < columnConfig.length; i++)
+			if (columnConfig[i].columnField == IUColumnConfig.COLUMN_ID) {
+				showingId = true;
+				break;
+			}
 	}
 
 	public int compare(Viewer viewer, Object obj1, Object obj2) {
@@ -34,13 +51,22 @@ public class IUComparator extends ViewerComparator {
 
 		String key1, key2;
 		if (key == IU_NAME) {
-			// Compare the iu names in the default locale
+			// Compare the iu names in the default locale.
+			// If a name is not defined, we use blank if we know the id is shown in another
+			// column.  If the id is not shown elsewhere, then we are displaying it, so use
+			// the id instead.
 			key1 = IUPropertyUtils.getIUProperty(iu1, IInstallableUnit.PROP_NAME);
 			if (key1 == null)
-				key1 = ""; //$NON-NLS-1$
+				if (showingId)
+					key1 = ""; //$NON-NLS-1$
+				else
+					key1 = iu1.getId();
 			key2 = IUPropertyUtils.getIUProperty(iu2, IInstallableUnit.PROP_NAME);
 			if (key2 == null)
-				key2 = ""; //$NON-NLS-1$
+				if (showingId)
+					key2 = ""; //$NON-NLS-1$
+				else
+					key2 = iu2.getId();
 		} else {
 			key1 = iu1.getId();
 			key2 = iu2.getId();
